### ISSUE TITLE
Issue with Fl_Native_File_Chooser::directory()

### ISSUE THREAD

**Issue by pvrose** [fltk/fltk#122]
On 2020-08-04 08:54, 'Philip Rose' via fltk.general wrote:
> I am using Fl_Native_File_Chooser, and calling directory to set the initial search directory does not seem to do so.

    Hmm, can confirm with the test/native_filechooser.exe.

    This used to work, I know it did. Must be a regression or change in windows API, not sure.
    I suggest opening an issue for this one.

-- 

I am using Fl_Native_File_Chooser, and calling directory to set the initial search directory does not seem to do so. My code is: [EDIT: I edited this comment to 'quote' your code so it displays indenting correctly on github. -erco]

 ```

   // Callback to open a single file chooser from a button
    static void cb_bn_browsefile(Fl_Widget* w, void* v) {
        browser_data_t* data = (browser_data_t*)v;
        // Open file chooser with supplied data
        Fl_Native_File_Chooser* chooser = new Fl_Native_File_Chooser(Fl_Native_File_Chooser::BROWSE_FILE);
        chooser->title(data->message.c_str());
        chooser->filter(data->pattern.c_str());
        size_t pos = data->filename->find_last_of("/\\");
        if (pos != string::npos) {
            string dir_name = data->filename->substr(0, pos + 1);
            chooser->directory(dir_name.c_str());
        }
        switch (chooser->show()) {
        case 0:
            // If 1 file is selected - write filename to input widget
            *data->filename = string(chooser->filename());
            (data->input_w)->value(data->filename->c_str());
            if (data->enable != nullptr) {
                // If a check widget and bool is supplied, set it
                *(data->enable) = true;
                data->enable_w->value(true);
            }
            break;
        case -1:
            // Error from dialog
            printf("ERROR: %s", chooser->errmsg());
            break;
        case 1:
            // Ignore user Cancel
            break;
        }
    }
```

This is a callback intended to be used with a generic browse button. browser_data_t is a structure that allows me to provide a variable for the filename, filter, title of the browser and Fl_Input to take the selected filename. I've tried to set directory to the full filename, just the directory part (with or without the final back/fore-slash). It still defaults to the last directory selected in the chooser. I am using W10, but with legacy file choosers. The supplied filename is usually the full pathname of the file including disc letter.

Regards Phil.

**Comment by erco77** [comment]
Adding my comments recently posted to fltk.general (before I knew about this issue):
```
	FWIW, seems to work in the test program /if/ I supply an absolute path
	to a file that exists:

	C:\Windows\System32\drivers\gmreadme.txt

	If I do that, the dialog opens in the drivers directory, and shows "gmreadme"
	in the input field (without the extension). Oddly it leaves the file in the
	list *unselected* -- why, I'm not sure.

	But if I give it a directory, it seems to only show the directory one up from
	the directory I supply, e.g. supplying either of these:

		c:\windows\system32
		c:\windows\system32\

	..shows the contents of the "Windows" directory (?), with "system32" in the
	input field. Why it doesn't show the contents of the system32 directory instead
	I'm not sure.

* * *

	Internally, at the WIN32 level, it's passing the directory pathname to the
	WIN32 OPENFILENAMEA structure as the value "_ofn_ptr->lpstrInitialDir", and
	then calls GetOpenFileNameW() to actually open the dialog.

	I'm sure there may be newer dialog techniques, but we're trying to support
	all platforms, not just the most recent. This feature should work.

	Here's the docs for the "ofn" (OPENFILENAMEA) structure's contents:
	https://docs.microsoft.com/en-us/windows/win32/api/commdlg/ns-commdlg-openfilenamea

	"Flags" we set are:

  _ofn_ptr->Flags |= OFN_NOVALIDATE;    // prevent disabling of front slashes
  _ofn_ptr->Flags |= OFN_HIDEREADONLY;  // hide goofy readonly flag
  // USE NEW BROWSER
  _ofn_ptr->Flags |= OFN_EXPLORER;      // use newer explorer windows
  _ofn_ptr->Flags |= OFN_ENABLESIZING;  // allow window to be resized (hey, why not?)
  _ofn_ptr->Flags |= OFN_NOCHANGEDIR;   // XXX: docs say ineffective on XP/2K/NT, but set it anyway..

	There are maybe other flags available to us now we should be setting.
	But I looked over the list of supported Flag values at the link above,
	and didn't see anything relevant to this issue.

	There's a lot of caveats for lpstrInitialDir that differ across platforms.
	It says for pre-Windows7, the behavior is:

"""
Windows 2000/XP/Vista:

    1. If lpstrFile contains a path, that path is the initial directory.
    2. Otherwise, lpstrInitialDir specifies the initial directory.
    3. Otherwise, if the application has used an Open or Save As dialog box
       in the past, the path most recently used is selected as the initial directory.
       However, if an application is not run for a long time, its saved selected path
       is discarded.
    [..etc].
"""

	I take it since we leave lpstrFile set to NULL #1 is skipped,
	so #2 takes effect on those older platforms.

	For Windows 7 (and presumably "up"), it says:

"""
Windows 7:

    1. If lpstrInitialDir has the same value as was passed the first time the application
       used an Open or Save As dialog box, the path most recently selected by the user
       is used as the initial directory.
    2. Otherwise, if lpstrFile contains a path, that path is the initial directory.
    3. Otherwise, if lpstrInitialDir is not NULL, it specifies the initial directory.
    4. If lpstrInitialDir is NULL and the current directory contains any files of the
       specified filter types, the initial directory is the current directory.
"""

	Assuming the app has not opened a dialog yet, we can rule out #1,
	and we currently leave lpstrFile unset (NULL) which should rule out #2,
	leaving #3 which /should/ be setting the initial directory.

	Which it does, sort of. (Kinda weird behavior I detailed above).

	I'm not really sure what to make of this offhand.
	It either involves some deeper parsing of that structure's docs,
	or it's just the usual Microsoft "no one knows for sure" behavior.

	I'm definitely no expert on the WIN32 native API, so if anyone sees
	a clear solution, feel free to weigh in.
```

**Comment by erco77** [comment]
Of interest, the attached code is my quick + dirty reduction of Fl_Native_File_Chooser_WIN32.cxx
as a pure win32 C++ program (no FLTK) that compiles and runs under mingw64. Builds with:

    g++ getopenfilename-dialog.cpp -o getopenfilename-dialog.exe -lcomdlg32

It runs as expected, browsing the contents of `C:\windows\system32\` directory as it's hard wired to do.
Which could mean we're doing something subtle in the FLTK code that might be breaking this.

Unfortunately I have be offline for a few days, so I can't continue to look into this during that time,
but perhaps some of you brave folks can determine what the issue is, otherwise I'll try to remember
to look into this on my return..
[getopenfilename-dialog.cpp.txt](https://github.com/fltk/fltk/files/5026110/getopenfilename-dialog.cpp.txt)

**Comment by fire-eggs** [comment]
I have been unable to reproduce this issue using a tweaked version of native_filechooser.cxx. The native file dialog behaved correctly for me on both Windows 7 and Win10.1809.

I would like to point out that the Windows dialog box will not accept a path using slashes! The initial directory string _must_ use backslashes ("C:\windows\Fonts" should work; "C:/windows32\Fonts" will not).

**Comment by pvrose** [comment]
Thanks Kevin,

Yes that is probably the problem. I am using a filename with foreslashes. I’ll test this tomorrow.

Regards Phil.

Sent from Mail for Windows 10

**Comment by fire-eggs** [comment]
>Yes that is probably the problem. I am using a filename with foreslashes.

Thanks for the update!!!

Then I feel the issue is more that Fl_Native_File_Chooser should ensure the path format is correct as required by Windows. This is ideally a detail the library should take care of, not the application programmer.

[But others may disagree ...]

**Comment by erco77** [comment]
Requoting here from my post on fltk.general:

	FWIW the Fl_Native_File_Chooser_WIN32.cxx was originally designed to force
	all paths to be unix style, converting to windows before opening the dialog,
	then converting back on its return.

	This was done with two existing methods called `Win2Unix()` and `Unix2Windows(),`
	but currently their actual use by the code is commented out for some reason,
	perhaps because an app expecting to work with Windows paths might get confused
	by paths turning into unix style unconditionally.

	Perhaps that was the reason the Win2Unix() stuff was commented out.. if so,
	it could probably be a little smarter and detect if it's a windows path going in,
	and NOT convert it so it stays windows paths in and out. And only if a unix path
	goes in, it converts it back and forth, preserving the style.

This sugg wouldn't work though for paths that are a mix of fronts and backs, such as the one
cited above with `"C:/windows32\Fonts"`.

Anyway, this could be done, but would need some guidance on how to handle the oddball cases
like the above.

**Comment by pvrose** [comment]
The odd cases would be me. I can never remember which angle slash is Unix or Dos.On 6 Aug 2020 18:38, erco77 <notifications@github.com> wrote:
Requoting here from my post on fltk.general:
FWIW the Fl_Native_File_Chooser_WIN32.cxx was originally designed to force
all paths to be unix style, converting to windows before opening the dialog,
then converting back on its return.

This was done with two existing methods called `Win2Unix()` and `Unix2Windows(),`
but currently their actual use by the code is commented out for some reason,
perhaps because an app expecting to work with Windows paths might get confused
by paths turning into unix style unconditionally.

Perhaps that was the reason the Win2Unix() stuff was commented out.. if so,
it could probably be a little smarter and detect if it's a windows path going in,
and NOT convert it so it stays windows paths in and out. And only if a unix path
goes in, it converts it back and forth, preserving the style.

This sugg wouldn't work though for paths that are a mix of fronts and backs, such as the one
cited above with "C:/windows32\Fonts".
Anyway, this could be done, but would need some guidance on how to handle the oddball cases
like the above.

**Comment by erco77** [comment]
All: it's best to make replies through the github interface. It avoids the email inserts, and makes it easier to quote replies. (Hard to tell from the above post which is new text and which is quoted)

**Comment by erco77** [comment]
I'll see if I can pull together a patch that reinstates the Unix2Win()/Win2Unix() code, while attempting to preserve the slash style used on input to the result on output, either as a patch or fork: which would you prefer for testing?

**Comment by erco77** [comment]
OK, here's a git patch against 1.4.x current:
[native-win32-patch-v1.txt](https://github.com/fltk/fltk/files/5043649/native-win32-patch-v1.txt)

What this patch does: If the specified directory:

    ..is unix style (front slashes), it converts to windows for the dialog, then back to unix on return.
    ..is windows style (back slashes), it remains a windows path on return.
    ..has no slashes, native windows style paths will be returned

The same behavior should be the case for the preset_file().
This should work for both file and directory choices.

The test/native_filechooser code has also been modified to allow a path to be specified on the command line, e.g.

    native-filechooser.exe c:/temp/bla.txt

Let me know if this does what you expect for all cases.

**Comment by pvrose** [comment]
I didn't necessarily wanted the slashes preserved. It was my fault that I thought / and \ had the same meaning in all os's, because I couldn't be bothered to remember which was which. I have one path remembered that has both.Phil.On 7 Aug 2020 22:14, erco77 <notifications@github.com> wrote:
OK, here's a git patch against 1.4.x current:
native-win32-patch-v1.txt
What this patch does: If the specified directory:
..is unix style (front slashes), it converts to windows for the dialog, then back to unix on return.
..is windows style (back slashes), it remains a windows path on return.
..has no slashes, native windows style paths will be returned

The same behavior should be the case for the preset_file().
This should work for both file and directory choices.
The test/native_filechooser code has also been modified to allow a path to be specified on the command line, e.g.
native-filechooser.exe c:/temp/bla.txt

Let me know if this does what you expect for all cases.

**Comment by erco77** [comment]
OK, though I plan to apply this patch to ensure both path types go through OK, since the OS supports both path types, and I know in vfx customers often like to work with forward slash UNC paths on windows to ensure the same paths work in a mixed environment. So hopefully this patch won't negatively affect your use case; I don't think it will.

**Comment by erco77** [comment]
Applied fix as 0da41da, closing.
